Refactoring parsing options

gchoqueux opened the pull request:
## Description
Simplify parsing options construction: 
   - parsing options will have two properties `in` and `out`
     - `in` will be data specification for **in data**.
     - `out` will be data specification for **out feature**.
  - in and out could be set, respectively, by `Source` and `Layer`.
```js 
// custom parsing  option to convert 'EPSG:4326' data to 'EPSG:4978' `Feature`.
const parsingOptions = { 
   in: { crs: 'EPSG:4326'  },
   out: { crs: 'EPSG:4978'  },
};
// For a layer the options will be
const parsingOptionsForLayer = { 
   in: layer.source,
   out: layer,
};
```

fix  #1480

gchoqueux committed:
fix(examples): update deprecated parsingOptions.

gchoqueux commented on examples/js/plugins/CSVnVRTParser.js:
```diff
@@ -152,7 +153,7 @@ var CSVnVRTParser = (function _() {
     function readLayer(layer, data, options, crs) {
         if (layer.OGRVRTLayer) {
-            var collection = new itowns.FeatureCollection(options.crsOut, options);
+            var collection = new itowns.FeatureCollection(options.out.crs, options.out);
```
**TODO**: just pass feature options needed

zarov commented on src/Core/Feature.js:
```diff
@@ -247,11 +247,12 @@ export default Feature;
 export class FeatureCollection {
     constructor(crs, options) {
         this.isFeatureCollection = true;
+        // is crs in ???
```
What do you mean by that ?

gchoqueux replied:
The `crs` parameter is always equal to `CRS.formatToEPSG(options.crs)`.
I think i will remove this parameter in next PR.

zarov commented on src/Layer/Layer.js:
```diff
@@ -222,7 +209,7 @@ class Layer extends THREE.EventDispatcher {
                 data = Promise.resolve(this.convert(feature, to));
             } else {
-                data = this.source.loadData(from, this.parsingOptions)
+                data = this.source.loadData(from, { in: this.source, out: this })
```
If `in` is always the source, why not set `options.in = this` in `Source#loadData` ?

gchoqueux replied:
yes

gchoqueux replied:
done

zarov commented on src/Layer/LabelLayer.js:
```diff
@@ -44,7 +44,9 @@ class LabelLayer extends Layer {
 
-        this.parsingOptions.buildExtent = true;
+        // Feature options ??
+        // move to layer...because generaly the Source is shared
```
Comment ?

gchoqueux replied:
In `View`, I add `labelLayer.buildExtent = true` when it's created by the option `layer.enableLayer` 
Because the features are shared between this `layer` and `labelLayer`.

zarov commented on src/Core/Deprecated/Undeprecator.js:
```diff
@@ -0,0 +1,29 @@
+export const deprecatedParsingOptionsToNewOne = (options) => {
+    /* istanbul ignore next */
+    if (options.crsOut) {
```
What about `crsIn` ?

gchoqueux replied:
yes, I will add a test on `crsIn`

zarov commented on src/Layer/Layer.js:
```diff
@@ -123,29 +123,16 @@ class Layer extends THREE.EventDispatcher {
-        this.parsingOptions = {
-            crsIn: this.source.crs,
-            overrideAltitudeInToZero: this.overrideAltitudeInToZero,
-            filter: this.filter || this.source.filter,
```
How do you handle this case ?

gchoqueux replied:
is handling in `Layer` class

zarov reviewed (changes requested):
This PR lacks some documentation in `Source` and `Layer` on what options can be passed. You should split the `GeoJsonParserOptions` in two maybe, and move it to them ?

gchoqueux commented:
> This PR lacks some documentation in `Source` and `Layer` on what options can be passed. You should split the `GeoJsonParserOptions` in two maybe, and move it to them ?

I add `GeoJsonParserOptions`  documentation. ~~, but I didn't split them.~~

I splitted the documentation but it isn't clear to link correctly new types.

zarov commented on src/Core/Feature.js:
```diff
@@ -247,11 +247,12 @@ export default Feature;
 export class FeatureCollection {
     constructor(crs, options) {
         this.isFeatureCollection = true;
+        // TODO: Replace crs parameter by CRS.formatToEPSG(options.crs)
```
Why not doing it now ?

gchoqueux replied:
I prefer in another PR

zarov commented on src/Parser/GeoJsonParser.js:
```diff
@@ -175,20 +176,30 @@ function jsonFeaturesToFeatures(crsIn, crsOut, jsonFeatures, filteringExtent, op
      * @property {boolean} [buildExtent=false] - If true the geometry will
-     * have an extent property containing the area covered by the geom
+     * have an extent property containing the area covered by the geometry
+     * @property {string} forceExtentCrs - force feature extent crs if buildExtent is true.
```
This should be _force**d**ExtentCrs_, because _force_ sounds like a boolean, while with _force**d**_ we know it is the CRS.

gchoqueux replied:
yes

gchoqueux replied:
done

zarov commented on src/Source/FileSource.js:
```diff
@@ -155,17 +157,19 @@ class FileSource extends Source {
         if (!features) {
-            options.buildExtent = true;
+            options.out.buildExtent = true;
+            options.out.forceExtentCrs = options.out.crs != 'EPSG:4978' ? options.out.crs : this.crs;
```
What about `this.crs == 'EPSG:4978'` ?

gchoqueux replied:
handled: 

is this.crs == 'EPSG:4978'  so buildExtent = false

zarov reviewed (commented):
I don't really know either, best option would be to put it in Feature.

gchoqueux committed:
refactoring(core): remove Layer.parsingOptions.

Layer.parsingOptions are removed.

Parser options parameter, with crsOut and crsIn, are deprecated.
Use new parsing options, '{ in, out }' Object, instead of.

New parsing options construction:

parsing options will have two properties in and out
* 'in' will be data specification for in data.
* 'out' will be data specification for out feature.

'in' and 'out' could be set, respectively, by 'Source' and 'Layer'.

// custom parsing  option to convert 'EPSG:4326' data to 'EPSG:4978' `Feature`.
const parsingOptions = {
   in: { crs: 'EPSG:4326'  },
   out: { crs: 'EPSG:4978'  },
};
// For a layer the options will be
const parsingOptionsForLayer = {
   in: layer.source,
   out: layer,
};

gchoqueux committed:
fix(doc): update deprecated parsingOptions.

gchoqueux committed:
fix(test): update deprecated parsingOptions.
